httpstan/views.py

```diff
@@ -558,3 +558,75 @@ async def handle_get_operation(request: aiohttp.web.Request) -> aiohttp.web.Resp
         message, status = f"Operation `{operation_name}` not found.", 404
         return aiohttp.web.json_response(_make_error(message, status=status), status=status)
     return aiohttp.web.json_response(operation)
+
+
+async def handle_log_prob(request: aiohttp.web.Request) -> aiohttp.web.Response:
+    """Calculate the log probability.
+
+    ---
+    post:
+      summary: Return the log probability of the unconstrained parameters.
+      description: >-
+        Returns the output of Stan C++ model class methods:
+        ``log_prob``.
+      consumes:
+        - application/json
+      produces:
+        - application/json
+      parameters:
+        - name: model_id
+          in: path
+          description: ID of Stan model to use
+          required: true
+          type: string
+        - in: body
+          name: data
+          description: >-
+              Data for the Stan Model.
+          required: true
+          schema: Data
+        - in: body
+          name: unconstrained_parameters
+          description: >-
+              Unconstrained parameters to calculate log probability for.
+          required: true
+          schema:
+            type: array
+        - in: body
+          name: adjust_transform
+          description: >-
+              Boolean to control whether we apply a Jacobian adjust transform.
+          required: false
+          schema:
+            type: boolean
+      responses:
+        "200":
+          description: Log probability of the unconstrained parameters.
+          schema:
+            type: number
```

Comment:
```suggestion
        "200":
          description: Log probability of the unconstrained parameters.
          schema:
            type: object
            properties:
              log_prob:
                type: number
                items: ...
```

I think the error comes from here.

Should it return an object? I'm not sure what should go in `items` if anything?

Reply:
@ahartikainen Thank you for the suggestions, I amended the first change. On this one, I don't think we need "items" given that the function returns a single value, however I now see the error below

```
Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 236, in _validate_openapi_spec_validator
    validate_v2_spec(self.specification)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/openapi_spec_validator/shortcuts.py", line 7, in validate
    return validator_callable(spec, spec_url=spec_url)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/openapi_spec_validator/validators.py", line 48, in validate
    raise err
openapi_spec_validator.exceptions.OpenAPIValidationError: {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas

Failed validating 'oneOf' in schema['properties']['paths']['patternProperties']['^/']['properties']['post']['properties']['responses']['patternProperties']['^([0-9]{3})$|^(default)$']:
    {'oneOf': [{'$ref': '#/definitions/response'},
               {'$ref': '#/definitions/jsonReference'}]}

On instance['paths']['/v1/models/{model_id}/log_prob']['post']['responses']['200']:
    {'description': 'Log probability of the unconstrained parameters.',
     'properties': {'log_prob': {'type': 'number'}},
     'type': 'object'}

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/apispec/utils.py", line 69, in validate_spec
    prance.BaseParser(spec_string=json.dumps(spec.to_dict()), **parser_kwargs)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 107, in __init__
    self.parse()
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 138, in parse
    self._validate()
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 173, in _validate
    validator(parsed)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/__init__.py", line 240, in _validate_openapi_spec_validator
    raise_from(ValidationError, v2_ex)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/prance/util/exceptions.py", line 25, in raise_from
    raise klass(*args) from from_value
prance.ValidationError: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/sgmcart3/Code/stan/httpstan/doc/conf.py", line 62, in make_openapi_spec
    fh.write(openapi.openapi_spec().to_yaml())
  File "/home/sgmcart3/Code/stan/httpstan/httpstan/openapi.py", line 50, in openapi_spec
    apispec.utils.validate_spec(spec)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/apispec/utils.py", line 71, in validate_spec
    raise exceptions.OpenAPIError(*err.args)
apispec.exceptions.OpenAPIError: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/cmd/build.py", line 279, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/application.py", line 278, in __init__
    self._init_builder()
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/application.py", line 337, in _init_builder
    self.events.emit('builder-inited')
  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/sphinx/events.py", line 118, in emit
    (listener.handler, name), exc) from exc
sphinx.errors.ExtensionError: Handler <function make_openapi_spec at 0x7f5d4df3b4d0> for event 'builder-inited' threw an exception (exception: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None))

Extension error:
Handler <function make_openapi_spec at 0x7f5d4df3b4d0> for event 'builder-inited' threw an exception (exception: ("{'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}} is not valid under any of the given schemas", 'oneOf', deque(['paths', '/v1/models/{model_id}/log_prob', 'post', 'responses', '200']), None, [<ValidationError: "'properties', 'type' do not match any of the regexes: '^x-'">, <ValidationError: "'$ref' is a required property">, <ValidationError: "Additional properties are not allowed ('type', 'properties', 'description' were unexpected)">], [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}], {'description': 'Log probability of the unconstrained parameters.', 'type': 'object', 'properties': {'log_prob': {'type': 'number'}}}, {'oneOf': [{'$ref': '#/definitions/response'}, {'$ref': '#/definitions/jsonReference'}]}, deque(['properties', 'paths', 'patternProperties', '^/', 'properties', 'post', 'properties', 'responses', 'patternProperties', '^([0-9]{3})$|^(default)$', 'oneOf']), None))
```

Reply:
I wish one didn't have to learn a new "language" (i.e., REST-based APIs) to write code. Alas, it seems like there's no alternative.

I *think* the convention is that you always encapsulate things in a JSON object. I'd have to do more searching to verify this. (It might even be a hard requirement.)

How about ``{"log_prob": value}`` or ``{"log_prob": value}``

Reply:
``{"data": value}`` might also work. Check out https://google.github.io/styleguide/jsoncstyleguide.xml for ideas. I think that document might also explain the encapsulation issue.

Reply:
Thank you. The sphinx module runs without any errors when removing the following:

```
        - in: body
          name: unconstrained_parameters
          description: >-
              Unconstrained parameters to calculate log probability for.
          required: true
          schema:
            type: array
        - in: body
          name: adjust_transform
          description: >-
              Boolean to control whether we apply a Jacobian adjust transform.
          required: false
          schema:
            type: boolean
```

Reply:
The issue seems to be regarding how I define `type: array`, replacing array with, say, boolean means that sphinx runs with no error.